Blazor RCL support and shared MAUI services

Eilon opened the pull request:
Several major changes in this PR:
- BlazorWebView now uses the existing IServiceProvider that is available in MAUI
	- You must enable MS.Extensions.DI to use it
	- @PureWeen - could you review this part of the change (fairly small)

- BlazorWebView initialization now typically occurs before the first RootComponent is added, so it uses change notification to add/remove RootComponents as the list changes
   - @SteveSandersonMS - this should look familiar to you because the code is now mostly copied from the WPF/WinForms versions for listening to root component changes
- Razor Class Library (RCL) static assets support
	- The NUPKG contains build props/targets that use the Static Web Assets feature of Razor to gather and package the static assets
	- The targets take all the Razor Static Web Assets and embed them (yuck) in the project assembly as EmbeddedResources, along with an EmbeddedManfiestFileProvider manifest (to preserve folder structure and filenames)
	- Platforms:
		- Windows (WinUI) is **fully implemented**
		- Android is **fully implemented**
		- iOS/MacCatalyst is **fully implemented**
	- @javiercn - this part is for you

- The new targets also include some workarounds for various issues:
	- Work around Razor components ("Content") being detected by the iOS/MacCatalyst SDK as BundleResources (workaround: Mark Razor components as RazorComponent)
	- Work around static assets ("Content") being detected by the iOS/MacCatalyst SDK as BundleResources (workaround: Temporarily remove wwwroot assets from Content, then add them back later)
	- @javiercn - this part is also for you

Eilon committed:
Fix static assets and RCL support for BlazorWebView
- Shared services between MAUI and Blazor
- Move common handler property code to common handler file

rmarinho committed:
Fix yaml triggers

rmarinho committed:
Run pr's for release branches

github-actions[bot] committed:
Automated dotnet-format update (#839)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rmarinho committed:
[Housekeeping] Add provisionating xcode to net6 (#836)

* Add provisionating xcode to net6

* Fix essentials device tests

* try again

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/xamarin/xamarin-android build main-a2156d41275b92de3cf23f7c79801327c74b7fe0-1 (#846)

Microsoft.Android.Sdk.Windows
 From Version 11.0.200-ci.main.226 -> To Version 11.0.200-ci.main.234

Dependency coherency updates

Microsoft.Dotnet.Sdk.Internal
 From Version 6.0.100-preview.4.21215.1 -> To Version 6.0.100-preview.4.21221.10 (parent: Microsoft.Android.Sdk.Windows

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>

PureWeen committed:
WinUI Navigation Handler (#801)

* WinUI Navigation Handler

* - PR comments

* - consolidate a few more APIs

* - fix namespaces

* - fix namespaces

* Update src/Compatibility/Core/src/WinUI/FlyoutPageRenderer.cs

Co-authored-by: Matthew Leibowitz <mattleibow@live.com>

Co-authored-by: Matthew Leibowitz <mattleibow@live.com>

mattleibow committed:
Improve Hot Reload Integration (#851)

Clancey committed:
Fixes HotReload (#853)

Reverts changed from: https://github.com/dotnet/maui/commit/7d4d1ce4b78ec941fe56126d9eaaee5fb69ff1b6#diff-2b4bb95e12307423f7126d91b43a1511b68d00535e9f2b5f535af85d7a388333R36

PureWeen committed:
Android Push/Pop Navigation (#837)

* Android Navigation

* - fix layout

* - changes

* - yay pushing

* - wire up nav to use bundle ids

* - cleanup

* - setup pop

* - build up nav stack

* - remove nav push

* - remove graph xml

* Update global.json

* - add android to non net6 controls project

* - fix namespace

* Update Microsoft.Maui-net6.sln

PureWeen committed:
iOS NavigationPageHandler (#852)

* iOS Navigation

* - wire up iOS Navigation

* - remove VET

* - rework with HR

* - fix namespace

* - fix HR

* - fix hr

* - fix SO exception

* - maybe?

* - nullabel fix

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/xamarin/xamarin-macios build 20210426.5 (#847)

Microsoft.MacCatalyst.Sdk , Microsoft.iOS.Sdk
 From Version 14.5.100-ci.main.620 -> To Version 14.5.100-ci.main.658

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>
Co-authored-by: Rui Marinho <me@ruimarinho.net>

github-actions[bot] committed:
Automated dotnet-format update (#854)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

jsuarezruiz committed:
Implement PickerHandler in WinUI (#779)

* Implement PickerHandler in WinUI

* Enab le nullable in some classes

* Fix build error

* Register MauiComboBox ResourceDictionary

* Updated PickerExtensions

* Updated MauiComboBox

Co-authored-by: Rachel Kang <rachelkang@microsoft.com>

rmarinho committed:
[Build] Update maestro with macOS and tvOS (#862)

* [Build] Update maestro with macOS and tvOS

* [Build] Add tvOS and macOS to workload/dogfood

* Add Microsoft.iOS.Windows.Sdk

Co-authored-by: Jonathan Peppers <jonathan.peppers@microsoft.com>

1d0n7kn0w committed:
Update README.md (#865)

PureWeen committed:
Merge branch 'release/6.0.1xx-preview4'

# Conflicts:
#	eng/Version.Details.xml
#	eng/Versions.props
#	src/Controls/samples/Controls.Sample/Pages/MainPage.cs
#	src/Controls/src/Core/AppHostBuilderExtensions.cs
#	src/Core/src/TaskExtensions.cs

Clancey committed:
ContainerViewControllers should use the correct background color. (#868)

Also set the title of the VC from a page

PureWeen committed:
Merge branch 'release/6.0.1xx-preview4'

github-actions[bot] committed:
Automated dotnet-format update (#872)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

eerhardt committed:
Fix CodeDom reference (#888)

I noticed that in my Maui application System.CodeDom was being referenced. This is because we don't have a `PrivateAssets="all"` on the PackageReference in Microsoft.Maui.Controls.Build.Tasks.csproj.

Adding the PrivateAssets="all", and removing unnecessary Conditions since this project only builds for netstandard2.0.

mattleibow committed:
ImageSource and ImageSourceServices (#759)

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/xamarin/xamarin-macios build 20210503.10 (#919)

Microsoft.macOS.Sdk , Microsoft.MacCatalyst.Sdk , Microsoft.iOS.Sdk , Microsoft.tvOS.Sdk
 From Version 11.3.100-ci.main.664 -> To Version 11.3.100-ci.main.697

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>

rmarinho committed:
Sign release branches by default (#921)

PureWeen committed:
Merge branch 'release/6.0.1xx-preview4'

# Conflicts:
#	.nuspec/Microsoft.Maui.Controls.MultiTargeting.targets
#	eng/Version.Details.xml
#	eng/Versions.props
#	src/Controls/src/Core/HandlerImpl/Window.Impl.cs
#	src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs
#	src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.cs
#	src/Controls/src/Core/Platform/Android/NavHostPageFragment.cs
#	src/Core/src/Core-net6.csproj
#	src/Core/src/Handlers/Picker/PickerHandler.Windows.cs
#	src/Core/src/Platform/Windows/PickerExtensions.cs
#	src/Core/src/TaskExtensions.cs

PureWeen committed:
Fix Main Merge (#929)

* Fix merge

* - fix missing catalyst apis

* - fix

* - remove catalyst checks

SteveSandersonMS commented on src/BlazorWebView/src/core/Android/AndroidAssetFileProvider.cs:
```diff
@@ -0,0 +1,201 @@
+						using (var file = destination.OpenRead())
+						{
+							var buffer = new byte[destination.Length];
+							file.Read(buffer, 0, (int)destination.Length);
```
Does this read all the existing static assets into memory on every app startup? If so it seems like it might be a bit inefficient and contribute to delaying the app startup, particularly for any apps that have any large static content items.

Not saying it should be blocking for a preview, but it might be worth tracking some plan to put an overall checksum in place - across all the static assets - and then just comparing that single top-level checksum so we know whether a new version has been installed since the last startup.

mattleibow replied:
Or maybe it is possible to just read directly from the assets files? There are some APIs and we could potentially just read the zip entries further.

Eilon replied:
Discussed with @Redth and @javiercn this morning and we will use a different approach for static assets using `MauiAssets`. This will all go away.

Eilon replied:
I'm putting the Android work on hold for now because it's not a high priority for this preview. I'll leave this code here anyway.

SteveSandersonMS commented on src/BlazorWebView/src/core/Android/BlazorWebViewHandler.Android.cs:
```diff
@@ -54,13 +56,42 @@ protected override void DisconnectHandler(AWebView nativeView)
+							}
+						});
+					}
+					_rootComponents.CollectionChanged += OnRootComponentsCollectionChanged;
```
There might be a race condition here. If L86 runs before the dispatcher continuation from L80, and the collection changes in between those two executions, then I think a root component would be added twice.

In practice it probably wouldn't happen because we're probably already on the dispatcher thread, but just in case, maybe it's worth *not* dispatching, and instead using `Dispatcher.CheckAccess()` (throwing if false) to be sure we didn't need to.

Eilon replied:
I think I've fixed this now (but the code is in another file now). But if equality works the way we think, I think the logic in the CollectionChanged handler would avoid any potential problem anyway, no?

SteveSandersonMS commented on src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs:
```diff
@@ -30,13 +32,42 @@ protected override void DisconnectHandler(WebView2Control nativeView)
+							}
+						});
+					}
+					_rootComponents.CollectionChanged += OnRootComponentsCollectionChanged;
```
Same potential race condition here

SteveSandersonMS commented on src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs:
```diff
@@ -102,13 +103,42 @@ protected override void DisconnectHandler(WKWebView nativeView)
+							}
+						});
+					}
+					_rootComponents.CollectionChanged += OnRootComponentsCollectionChanged;
```
Also race here (or again, maybe not, depending on whether we're already on the dispatcher)

SteveSandersonMS commented on src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs:
```diff
@@ -153,10 +177,28 @@ public static void MapRootComponents(BlazorWebViewHandler handler, IBlazorWebVie
+						await item.RemoveFromWebViewManagerAsync(_webviewManager);
+					}
+				});
+			}
```
The amount of duplication across the three hosting flavours here makes me wonder if something common could be extracted from all three. Not suggesting you should hold up this PR for that, but might save time in the long run.

Eilon replied:
I've cleaned it up and put the common code in a common file.

SteveSandersonMS commented on src/Controls/samples/Controls.Sample.SingleProject/Startup.cs:
```diff
@@ -1,18 +1,43 @@
 				.UseMauiApp<MyApp>();
+
+			if (UseBlazor)
+            {
```
There are some pretty eclectic variations in indentation across different files in this PR. Is it because the MAUI repo as a whole has differing conventions in different places?

But this `{` in particular is a true maverick.

Eilon replied:
The MAUI bot will fix my mess later 😄

SteveSandersonMS commented on src/Controls/samples/Controls.Sample.WinUI/Maui.Controls.Sample.WinUI.csproj:
```diff
@@ -21,4 +24,7 @@
   </ItemGroup>
+
+  <!-- Normally this import would be automatic through the NuGet package. We do it manually for projects inside this solution. -->
+  <Import Project="..\..\..\BlazorWebView\src\core\build\net6.0-windows\Microsoft.AspNetCore.Components.WebView.Maui.targets" Condition="$(TargetFramework.StartsWith('net6.0-windows'))" /> 
```
Rather than adding these in different places manually, would it work to add `Directory.Build.props` and `Directory.Build.targets` files?

Also is there any symbol that could be used in place of the `net6.0-windows`? We know for sure that version number is going to get stale.

SteveSandersonMS commented on src/Controls/samples/Controls.Sample/Maui.Controls.Sample-net6.csproj:
```diff
@@ -43,21 +43,20 @@
+    <Content Update="wwwroot\index.html">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
```
Should these be here? Doesn't the SDK do this automatically for files in `wwwroot`?

SteveSandersonMS commented on src/Controls/samples/Controls.Sample/Startup.cs:
```diff
@@ -1,16 +1,21 @@
-﻿using System;
+﻿#if NET6_0_OR_GREATER
+#define BLAZOR_ENABLED
+#endif
```
I must be missing something here. When would `NET6_0_OR_GREATER` ever not be defined? That's all we support here, isn't it?

rmarinho replied:
we build this sln in a not .net6 world also for now (legacy to better have tool support)

Eilon replied:
Right, this shared sample is currently used also in non-.NET 6 scenarios, so all Blazor stuff is excluded from such builds. I think this eventually goes away and it's all .NET 6, but for now it needs to stay here to compile.

github-actions[bot] committed:
Automated dotnet-format update (#935)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

SteveSandersonMS commented on src/Controls/samples/Controls.Sample/Startup.cs:
```diff
@@ -59,23 +64,36 @@ public void Configure(IAppHostBuilder appBuilder)
+#if BLAZOR_ENABLED
 								typeof(BlazorPage),
 #else
 								throw new NotSupportedException("Blazor requires .NET 6 or higher."),
```
It will be interested to see the project template. I deeply hope it doesn't look anything like this file, which is mystifying.

Eilon replied:
The project templates will be .NET 6 only so all the `#if`s go away.

javiercn commented on src/Controls/samples/Controls.Sample/wwwroot/index.html:
```diff
@@ -6,6 +6,7 @@
     <title>Blazor Android app</title>
     <base href="/" />
     <link href="css/app.css" rel="stylesheet" />
+    <link href="_content/MauiRazorClassLibrarySample/MauiRazorClassLibrarySample.bundle.scp.css" rel="stylesheet" />
```
This is wrong I believe. You should reference `Controls.Sample.styles.css`, the `.bundle.scp.css` files only contains the bundle for the scoped css files, it doesn't have import references to other libraries. This works right now because you don't have transitive RCLs in the app.

Eilon replied:
For others watching: I met with @javiercn and @Redth today and we have a different solution that should fix this.

javiercn commented on src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:
```diff
@@ -1,11 +1,12 @@
     <RootNamespace>Maui.Controls.Sample.Droid</RootNamespace>
     <AssemblyName>Maui.Controls.Sample.Droid</AssemblyName>
     <IsPackable>false</IsPackable>
+    <StaticWebAssetsEnabled>true</StaticWebAssetsEnabled>
```
Hmm, not sure I understand why this is needed, it should be enabled based on the fact you are referencing the SDK

javiercn commented on src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:
```diff
@@ -15,4 +16,30 @@
     <PackageReference Include="Xamarin.AndroidX.Browser" />
   </ItemGroup>
 
+  <Target Name="_StaticWebAssetsComputeFilesToBuild" AfterTargets="GetCopyToOutputDirectoryItems" DependsOnTargets="ResolveStaticWebAssetsInputs">
+    <ItemGroup>
+      <_ExternalBuildStaticWebAsset Include="%(StaticWebAsset.FullPath)" Condition="'%(SourceType)' != ''">
+        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        <TargetPath>$([MSBuild]::MakeRelative('$(MSBuildProjectDirectory)','$([MSBuild]::NormalizePath('wwwroot\%(BasePath)\%(RelativePath)'))'))</TargetPath>
+      </_ExternalBuildStaticWebAsset>
+      <_SourceItemsToCopyToOutputDirectory Remove="@(_ExternalBuildStaticWebAsset)" />
+      <_SourceItemsToCopyToOutputDirectory Include="@(_ExternalBuildStaticWebAsset)" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="ZipWWWRootPath" AfterTargets="_FindJavaStubFiles" BeforeTargets="_CreateBaseApk" DependsOnTargets="_StaticWebAssetsComputeFilesToBuild">
+    <MakeDir Directories="$(ProjectDir)Assets" />
+    <ZipDirectory SourceDirectory="$(ProjectDir)$(OutputPath)wwwroot" DestinationFile="$(ProjectDir)Assets\wwwroot.zip" Overwrite="true" />
+  </Target>
+
+  <Target Name="_StaticWebAssetsUpdateAssetPaths" AfterTargets="_FindJavaStubFiles" BeforeTargets="_CreateBaseApk" DependsOnTargets="ZipWWWRootPath">
+    <ItemGroup>
+      <WWWRootZipStaticWebAsset Include="Assets\wwwroot.zip" />
+      <AndroidAsset Include="@(WWWRootZipStaticWebAsset)" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="DeleteAsset" AfterTargets="IncrementalClean" DependsOnTargets="ZipWWWRootPath">
+    <Delete Files="$(ProjectDir)Assets\wwwroot.zip" />
+  </Target>
```
Is this something that will appear on the template (almost sure the answer is no, but have to ask 😄 )

Eilon replied:
It won't 😄

javiercn commented on src/BlazorWebView/src/core/iOS/BlazorWebViewHandler.iOS.cs:
```diff
@@ -153,10 +177,28 @@ public static void MapRootComponents(BlazorWebViewHandler handler, IBlazorWebVie
+					var newItems = eventArgs.NewItems!.Cast<RootComponent>();
+					var oldItems = eventArgs.OldItems!.Cast<RootComponent>();
+
+					foreach (var item in newItems.Except(oldItems))
```
Can this be a problem here? I'm not sure if we've defined equality for the root component type.

SteveSandersonMS replied:
I would have thought the default struct equality rules would have the right behavior in this case.

Eilon replied:
It's the same exact code from WPF/WinForms. I'll log an issue to confirm.

javiercn commented on src/BlazorWebView/src/core/Windows/BlazorWebViewHandler.Windows.cs:
```diff
@@ -45,17 +76,20 @@ private void StartWebViewCoreIfPossible()
+			// It is critical to avoid calling System.IO.Path APIs that use the "current directory"
+			// when they are passed in a relative path, because the app has no control over where
+			// the "current directory" is (it could be anywhere).
+			var entryAssemblyPath = global::System.Reflection.Assembly.GetEntryAssembly()!.Location;
```
What's the reason for using `global::` here?

Eilon replied:
To avoid a conflict with some sub-namespace that has `System` in it. But I think this code is gone now.

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/xamarin/xamarin-android build main-c60635bf2161f6e3b2d74ebcd7a3f96deef956a2-1 (#918)

Microsoft.Android.Sdk.Windows
 From Version 11.0.200-ci.main.238 -> To Version 11.0.200-ci.main.250

Dependency coherency updates

Microsoft.Dotnet.Sdk.Internal
 From Version 6.0.100-preview.4.21221.10 -> To Version 6.0.100-preview.5.21229.1 (parent: Microsoft.Android.Sdk.Windows

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>
Co-authored-by: Rui Marinho <me@ruimarinho.net>

javiercn reviewed (commented):
I've made a few comments, things look good in general, just small fixes.

PureWeen committed:
Merge branch 'release/6.0.1xx-preview4'

# Conflicts:
#	src/Controls/src/Core/Controls.Core-net6.csproj
#	src/Controls/src/Core/Handlers/NavigationPage/NavigationPageHandler.Android.cs
#	src/Controls/src/Core/Platform/Android/NavHostPageFragment.cs

PureWeen committed:
Fix main merge

Eilon commented on src/BlazorWebView/src/core/MauiAssetFileProvider.Windows.cs:
```diff
@@ -0,0 +1,86 @@
+		IFileInfo? PlatformGetFileInfo(string subpath)
+		{
+			// TODO: HACK: For now we strip out the folder because it is not preserved in the WinUI assets. We need to address this in the MauiAsset targets.
+			var fileWithoutFolders = Path.GetFileName(subpath);
```
@Redth - FYI the current Resizetizer stuff seems to remove folder paths, so if I have a file `wwwroot\css\main\foo.css` the file ends up in `Appx\Assets\foo.css` and I lose the sub-folders. We need a way to preserve the folder structure.

(I tried hacking up some MSBuild goop to do it but couldn't figure it out.)

Do you have any concerns with adding support for sub-folders in MauiAssets?

Eilon commented on src/BlazorWebView/src/core/MauiAssetFileProvider.Windows.cs:
```diff
@@ -0,0 +1,86 @@
+			// TODO: HACK: For now we strip out the folder because it is not preserved in the WinUI assets. We need to address this in the MauiAsset targets.
+			var fileWithoutFolders = Path.GetFileName(subpath);
+
+			// TODO: Also, instead of using WinUI Package APIs, we're going directly to the filesystem. They all seem to throw. Maybe we're on the wrong thread?
```
@Redth - also FYI on this. Using the package APIs, which are all async, e.g. `GetItemAsync()`, would all throw mysterious C++ exceptions. I wonder if I'm running it on the wrong thread? (Which would be annoying, because nothing else here is async, and calling `GetResult()` is awful.)

Did you ever see any exceptions while using those APIs?

Eilon commented on src/BlazorWebView/src/core/build/Microsoft.AspNetCore.Components.WebView.Maui.targets:
```diff
@@ -0,0 +1,6 @@
+﻿<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="ConvertStaticWebAssetsToMauiAssets" BeforeTargets="ResizetizeCollectItems">
+    <!-- TODO: Implement this -->
+    <Message Text="Converting assets..." />
```
@javiercn - do your StaticWebAsset magic here. I believe it's the right place.

Eilon commented:
For anyone looking, I'm focusing only on WinUI working right now. Once that's ship-shape it'll make bringing up other platforms much easier because almost all the static asset code will be platform-independent shared code. Only the MauiAssetFileProvider will be platform-dependent.

github-actions[bot] committed:
Automated dotnet-format update (#941)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Clancey committed:
Add a new Attribute for static methods. [OnHotReload] (#833)

* Add a new Attribute for static methods. [OnHotReload]

* Fix Nullability

PureWeen committed:
Merge branch 'release/6.0.1xx-preview4'

rmarinho committed:
Onboard with OneLocBuild Task and Add Security checks (#631)

* Initial commit to move to one loc

* Fix templates reference

* Don't depend on validate

* try fake it

* Use loc branch variable

* One loc

* try fix path

* Fix StringResources filename

* Fix git add

* Don't use variable

* remove xliff

* test main

* Set main again

* Simplify policheck compliance

* Add loc to triggers

* Bump .NET version

* Update merge-translations-update.yml

* Update security_compliance.yml

rmarinho committed:
This should be ci.main

rmarinho committed:
[Oneloc]Fix loc build on CI

rmarinho committed:
Don't use LclFile

rmarinho committed:
Revert "Don't use LclFile"

This reverts commit c040e5130a1d2da44fe606e22f469deb7c2105e8.

mattleibow committed:
Allow derived image sources to indicate changes (#956)

jsuarezruiz committed:
Implement Completed event in EntryHandler (#820)

* Implement Completed in WinUI

* Implement Completed on Android and iOS EntryHandlers

* Added TODO comment

* Updated EntryStub

* Enable nullable in Windows EntryHandler

* Updated Windows EntryHandler

Co-authored-by: Rui Marinho <me@ruimarinho.net>

github-actions[bot] committed:
Automated dotnet-format update (#965)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

eerhardt committed:
Fix references of Essentials so unnecessary packages don't get referenced/downloaded (#954)

- System.ValueTuple is no longer necessary
- System.Numerics.Vectors is only necessary on netstandard, not on net6+.

rmarinho committed:
Only commit files under src

jsuarezruiz committed:
[Housekeeping] Remove AppKit references (#968)

* Remove AppKit references in Core

* Removed MaciOS folder

Eilon commented:
STATUS: I have confirmed that WinUI and Android are both functional. I'd like to get these checked in ASAP to unblock other work.

@javiercn / @SteveSandersonMS - please let me know if you have any concerns on the Blazor-isms.
@PureWeen - there's not much MAUI-isms here, but I'd appreciate a look all the same.

Clancey committed:
Handle Safe Area on iOS. It's an Opt in via Interface (#895)

* Handle Safe Area on iOS. It's an Opt in via Interface

* Added a check to SafeArea

github-actions[bot] committed:
Automated dotnet-format update (#973)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

vs-mobiletools-engineering-service2 committed:
[LOC_HB] string translations (#969)

Co-authored-by: VS Mobile Engineering Serice Account <vs-mobiletools-engineering-service2@microsoft.com>

Clancey committed:
Picker cleanup (#893)

* Core now compiles for iOS/Android with the new IPicker

* Controls now uses the new Interface

* Windows now uses the new IPicker

* Added IItemDelegate, IPicker now uses it.

* Fixing Casing

* temp

* Now fixed file casing

* Fixed Device Tests

* Added ItemDelegateList for performance, now using it on Windows

* Fixed the Build

* Update src/Core/tests/DeviceTests/Stubs/PickerStub.cs

* Update Picker.Impl.cs

* Bringing back the prettier code

Co-authored-by: Matthew Leibowitz <mattleibow@live.com>

jsuarezruiz committed:
Implement WinUI ProgressBarHandler native events (#874)

* Implement WinUI ProgressBarHandler native events

* Enable nullable in WinUI ProgressBarHandler

jsuarezruiz committed:
Implement native events in WinUI SearchBarHandler (#875)

* Implement native events in WinUI SearchBarHandler

* Enable nullable in WinUI SearchBarHandler

Co-authored-by: Rui Marinho <me@ruimarinho.net>

Clancey committed:
Font weight vs FontAttributes, Moved Named Sizes to Controls (#940)

vs-mobiletools-engineering-service2 committed:
[LOC_HB] string translations (#992)

Co-authored-by: VS Mobile Engineering Serice Account <vs-mobiletools-engineering-service2@microsoft.com>

vs-mobiletools-engineering-service2 committed:
[LOC_HB] string translations (#987)

Co-authored-by: VS Mobile Engineering Serice Account <vs-mobiletools-engineering-service2@microsoft.com>

github-actions[bot] committed:
Automated dotnet-format update (#983)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

jsuarezruiz committed:
Implement HorizontalTextAlignment property in WinUI Picker (#947)

* Implement HorizontalAlignment property in WinUI Picker

* Update package.ps1

* Update package.ps1

* Update PickerHandler.Windows.cs

Co-authored-by: Matthew Leibowitz <mattleibow@live.com>
Co-authored-by: Rui Marinho <me@ruimarinho.net>

jsuarezruiz committed:
Implement basic Editor properties in WinUI (#739)

* Implement basic Editor properties in WinUI

* Fix build error

* Fix build error

* Protect NRE in TextBoxExtensions

* Revert unnecessary changes

* Updated TextBoxExtensions

* Fix font reference in sample

Co-authored-by: Rachel Kang <rachelkang@microsoft.com>

rmarinho committed:
Merge branch 'release/6.0.1xx-preview4'

# Conflicts:
#	eng/Version.Details.xml
#	eng/Versions.props
#	eng/package.ps1
#	eng/pipelines/common/variables.yml
#	src/Compatibility/ControlGallery/src/WinUI/_13109IssueHelper.cs
#	src/Compatibility/Core/src/iOS/CADisplayLinkTicker.cs
#	src/Compatibility/Core/src/iOS/Cells/CellRenderer.cs
#	src/Compatibility/Core/src/iOS/Cells/ViewCellRenderer.cs
#	src/Compatibility/Core/src/iOS/Extensions/Extensions.cs
#	src/Compatibility/Core/src/iOS/Extensions/FlowDirectionExtensions.cs
#	src/Compatibility/Core/src/iOS/HandlerToRendererShim.cs
#	src/Compatibility/Core/src/iOS/IOSDeviceInfo.cs
#	src/Compatibility/Core/src/iOS/NativeBindingService.cs
#	src/Compatibility/Core/src/iOS/Renderers/CarouselPageRenderer.cs
#	src/Compatibility/Core/src/iOS/Renderers/ImageRenderer.cs
#	src/Compatibility/Core/src/iOS/Renderers/NavigationRenderer.cs
#	src/Controls/src/Xaml/Controls.Xaml-net6.csproj
#	src/Core/src/Handlers/Picker/PickerHandler.Standard.cs
#	src/Core/src/Platform/iOS/ColorExtensions.cs
#	src/Core/src/Platform/iOS/LayoutView.cs

rmarinho committed:
Merge branch 'main' of https://github.com/dotnet/maui

rmarinho committed:
Fix merge

Clancey committed:
Fixed android Fonts for older OS versions (#1001)

jsuarezruiz committed:
Fix WinUI font issue in TextBoxExtensions (#1005)

mattleibow committed:
Merge remote-tracking branch 'origin/release/6.0.1xx-preview4'

rmarinho committed:
Fix isLocPRBranch

Redth commented on src/Controls/samples/Controls.Sample.Droid/Maui.Controls.Sample.Droid-net6.csproj:
```diff
@@ -1,11 +1,12 @@
-<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk.Razor">
```
@jonathanpeppers how do we make this not needed, and instead import from a property?

jonathanpeppers replied:
So if we followed the pattern of what `$(UseMaui)` will eventually be, maybe we should come up with a new property like `$(UseRazor)`?

I think we could do something like:

```xml
<PropertyGroup>
  <_RazorSdkImportsMicrosoftNetSdk  Condition=" '$(UseRazor)' == 'true' ">false</_RazorSdkImportsMicrosoftNetSdk>
</PropertyGroup>
<Import Condition=" '$(UseRazor)' == 'true' " Sdk="Microsoft.NET.Sdk.Razor" Project="Sdk.props" />
```
And put this in `Microsoft.Maui.Controls.props` ^^

We'd also need to import in `Microsoft.Maui.Controls.targets`:

```xml
<Import Condition=" '$(UseRazor)' == 'true' " Sdk="Microsoft.NET.Sdk.Razor" Project="Sdk.targets" />
```

Then any project that has `Microsoft.NET.Sdk.Razor` would define `<UseRazor>true</UseRazor>` instead.

Eilon commented on src/Core/src/Hosting/AppHostBuilderExtensions.cs:
```diff
@@ -101,6 +102,22 @@ public static IAppHostBuilder UseMauiServiceProviderFactory(this IAppHostBuilder
 			return builder;
 		}
 
+		public static IAppHostBuilder UseMicrosoftExtensionsServiceProviderFactory(this IAppHostBuilder builder)
```
@PureWeen / @Redth / @rmarinho - any problem with adding this extension method? Hopefully we figure out a way to avoid this entirely, but for preview4, how's this?

github-actions[bot] committed:
Automated dotnet-format update (#1012)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

jsuarezruiz committed:
Allow pages without content (#1015)

Eilon commented:
OK folks, I think I finally have this stuff figured out. There's a bit of duct tape in there, but it's entirely hidden from the user's projects, meaning we can fix much of it post preview4 (and discussions are underway for many of those issues).

I tested in these scenarios:
- Project type: SingleProject vs. Multi-project
- Platform: WinUI, Android, iOS, MacCatalyst
- Asset location: All static assets/Razor files in main project vs. some assets/Razor files coming from Razor Class Libraries (RCL)

I didn't test every combinatorial scenario, but I did cover most.

I would appreciate a review ASAP (the code is much less than before!) so I can unblock the remaining Blazor work (update dependencies in #933, and then project templates).

Eilon commented:
Except one thing, I 100% messed up a rebase, so fixing that 😢

Eilon commented:
For numerous reasons, I'm closing this PR and I opened a new PR here: https://github.com/dotnet/maui/pull/1021
